### PR TITLE
Partial Setoids and Partial Equivalence Relations

### PR DESCRIPTION
**Pull request by TOTBWF**
This is a continuation of #842, and is meant to explore the impacts of adding PERs and Partial Setoids. 

Currently, this PR updates the definition of `IsEquivalenceRelation` from:
```
record IsEquivalence {A : Set a} (_≈_ : Rel A ℓ) : Set (a ⊔ ℓ) where
  field
    refl  : Reflexive _≈_
    sym   : Symmetric _≈_
    trans : Transitive _≈_
```

```
record IsEquivalence {A : Set a} (_≈_ : Rel A ℓ) : Set (a ⊔ ℓ) where
  field
    isPartialEquivalence : IsPartialEquivalence _≈_
    refl  : Reflexive _≈_
```
This makes logical sense, but is a major breaking change, even though solving the issues is mechanical. It would be possible to keep the definition of `IsEquivalence` the same, and "lower" it into a proof of `IsPartialEquivalence` by throwing away the `refl`, but this feels a bit clunky.

The things left to do on this are:
- [ ] generalize some proofs to work over PERS/partial setoids
- [ ] implement partial setoids where it makes sense
- [ ] explore a rework the algebraic hierarchy, with a focus on making it easy to work with algebraic substructures

### CODE REVIEW

**Comment by MatthewDaggitt** [comment]
Hi @TOTBWF, thanks for exploring this. I'm afraid though that I don't think this is viable to merge in, because as you identify, it's a major non-backwards compatible change. Personally speaking, it would break every piece of Agda code that I've ever written, and I'm sure I'm not alone in that! Possibly it's viable in some far-off mythical v2.0 but not anytime soon.

I think @JacquesCarette's other suggestion of a separate parallel hierarchy for partial equivalences is probably a better bet...

**Comment by TOTBWF** [comment]
Yeah, that does seem like the way to go. Lowering the equivalences shouldn't be _too_ onerous, and the only downside would be some weird scoping behaviour, but that is probably better than breaking existing code.

**Comment by JacquesCarette** [comment]
Thanks @TOTBWF - that was a lot of hard work to show the full impact of this, and it sure is appreciated! Too bad the outcome is that merging as first envisaged isn't viable.  I'm still eager to see the alternate hierarchy develop.

**Comment by HuStmpHrrr** [comment]
I suppose you can make `Relation.Binary.Reasoning.Base.Single` implemented by PER reasoning, so less code is duplicated.

**Comment by TOTBWF** [comment]
That makes more sense actually, then you could just apply `refl` in the other cases. Thanks 🙂

**Comment by HuStmpHrrr** [comment]
oh apologies if my suggestion was misleading. it is less good if you modify existing interfaces. this usually breaks dependencies. waht I was saying is that `Relation.Binary.Reasoning.Base` is a specialization of PER reasoning. for example, if you define `Relation.Binary.Reasoning.Partial`, which requires transitivity and symmetry, then `Relation.Binary.Reasoning.Base` merely requires one additional requirement, i.e. reflexivity. by having `Relation.Binary.Reasoning.Base` implemented in terms of `Relation.Binary.Reasoning.Partial`, you can reuse most of your code. then `PartialSetoid` reasoning is just a tedious realization of `Relation.Binary.Reasoning.Partial`.

**Comment by TOTBWF** [comment]
Ah, that makes far more sense, will make those changes

File: src/Relation/Binary/Reasoning/PartialSetoid.agda [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+open import Relation.Binary
+
+module Relation.Binary.Reasoning.PartialSetoid {s₁ s₂} {A : Set s₁} (S : PartialSetoid A s₂) where
+
+open PartialSetoid S
+
+------------------------------------------------------------------------
+-- Publicly re-export base contents
+
+open import Relation.Binary.Reasoning.Base.Partial _≈_ trans public
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Reasoning/PartialSetoid.agda]: You haven't actually included this file in the PR :smile:

File: src/Relation/Binary/Core.agda [hunk 2]
```diff
@@ -238,6 +238,17 @@ record NonEmpty {A : Set a} {B : Set b}
     {y}   : B
     proof : T x y
 
+------------------------------------------------------------------------
+-- Partial Equivalence relations
+--
+-- Equivalence relations are not defined in terms of their partial
+-- counterparts for backwards-compatibility reasons.
+
+record IsPartialEquivalence {A : Set a} (_≈_ : Rel A ℓ) : Set (a ⊔ ℓ) where
```
Comment (MatthewDaggitt) [inline src/Relation/Binary/Core.agda]: I don't think this needs to go in the `Core` module.

File: src/Relation/Binary.agda [hunk 3]
```diff
@@ -59,6 +59,25 @@ record Preorder c ℓ₁ ℓ₂ : Set (suc (c ⊔ ℓ₁ ⊔ ℓ₂)) where
...
   open IsPreorder isPreorder public
 
+------------------------------------------------------------------------
+-- Partial Setoids
+
+-- Partial equivalence relations are defined in Relation.Binary.Core.
+-- Furthermore, PartialSetoid is parameterized by a type to make
+-- dealing with subsets more feasible.
+
+record PartialSetoid {c} (Carrier : Set c) ℓ  : Set (suc (c ⊔ ℓ)) where
```
Comment (MatthewDaggitt) [inline src/Relation/Binary.agda]: I'm wondering if this and `IsPartialEquivalence` should go in a new module `Relation.Binary.Partial`? I'm not too keen on having anything with an explicit `Carrier` in alongside the rest of the hierarchy...

Comment (JacquesCarette) [reply src/Relation/Binary.agda]: I like that idea.

File: src/Relation/Binary/Reasoning/PartialSetoid.agda (hunk 1 above)
Comment (TOTBWF) [reply src/Relation/Binary/Reasoning/PartialSetoid.agda]: Just goes to show that I shouldn't be allowed to commit before I've had my coffee

**Comment by MatthewDaggitt** [comment]
Thanks for all the hardwork @TOTBWF! Clearly you've been up to a lot :+1: However it's a little difficult to review all of this at once. Do you think we could split this up into a few different pull requests?

To start with maybe the first one should just contain the following?
- Relation.Binary.Partial
- Relation.Binary.Partial.Reasoning

**Comment by TOTBWF** [comment]
Sorry for taking so long to respond, life decided to make me pretty busy this week.

That sounds reasonable to me. I've been toying around with this a bit more, and I _think_ I can pack away the carrier type for `PartialSetoid` and still get reasonable definitions for subsets/substructures/quotients. I'll open a new PR in the next couple of days for the basic definitions, and we can go from there.

As an aside, thank you for all of the commentary on this, it has been super helpful 🙂

**Comment by TOTBWF** [comment]
Closed in favor of #852

**Pull request closed without merging**